Added documentation for @typePolicy (#586)
apollographql/apollo-ios-dev: docs/source/caching/cache-key-resolution.mdx

> To learn more, read about how the cache [normalizes objects by cache key](./introduction#normalizing-objects-by-cache-key).

For most uses cases, you can configure cache keys declaratively using the methods described here. For advanced use cases not supported by declarative cache keys, you can configure cache keys [programmatically](./programmatic-cache-keys).

## Schema Extensions

> Declarative cache keys work by adding directives to your backend schema types to indicate how cache keys should be resolved for those times. To do this, you can extend your backend schema by creating a schema extension file with the file extension `.graphqls`. You can add any number of schema extension files to your project.

Author: Types. Fixing now.

> **Note:** All of a type's key fields must return a scalar type. To use non-scalar fields in cache keys, try using [programmatic cache key configuration](./programmatic-cache-keys).

### Multiple Key Fields

You can specify _multiple_ key fields for an object if they are all required to uniquely identify a particular cache entry. Multiple key fields should be seperated by a single space.

```graphql title="typePolicies.graphqls"
extend type Author @typePolicy(keyFields: "firstName lastName")
```

The resolved cache key will include all declared key fields separated by the `+` character. In this case, an `Author` object with a `firstName` of `"Ray"` and a last name of `"Bradbury"` would resolve to have a cache key of `Author:Ray+Bradbury`.

### Type policies on interface types

It is also possible to use `@typePolicy` on _interfaces_. This specifies a default set of key fields for all types that implement that interface.

```graphql title="extra.graphqls"
extend interface Node @typePolicy(keyFields: "id")
```

> If an object type has a type policy of its own, it will take precedence and the type policy on an interface it implements will not be applied.

Author: Done

## Caveats & limitations

Declarative cache key resolution has a few notable quirks and limitations you should be aware of while implementing your `@typePolicy` directives:

> 1. If any type implements two interfaces with conflicting type policies, a validation error will be thrown when executing code generation. This can be corrected by applying a type policy to the type directly to disambiguate the conflict.

Reviewer: During the review process for this feature, we decided against allowing types to overwrite an interface's typePolicy. If a type has a policy, all interfaces it inherits must have the exact same typePolicy (or no policy).

> 2. When an object is returned for a field of an interface type where the interface has a `@typePolicy`, Apollo iOS will first attempt to find a `@typePolicy` for the concrete type by using the `__typename` field of the returned object. If the type does not have an `@typePolicy` of its own, the interface's `@typePolicy` will be applied.

Author: This change is not an accurate description. The original text is correct.

> If the same object is returned for multiple fields of different interface types with conflicting type policies, it is possible that the same object is resolved with two different type policies, leading to cache inconsistencies.

Reviewer 2: This part feels like it needs its own subheader:

```suggestion
### Cache inconsistencies

If the same object is returned for fields of different interfaces with conflicting type policies, it might be resolved using two different policies. This can cause cache inconsistencies.
```

> In most circumstances, validation that is run during code generation prevents this from occuring. However this may still occur if changes are made to the schema after code generation runs (or after your application is published).

Reviewer 2: ```suggestion
In most circumstances, validation that is run during code generation prevents this from occurring. However, this may still occur if changes are made to the schema after code generation runs or after your application is published.
```

This is possible if:

>   A) A type is added to your schema after code generation is executed and that type implements two different interfaces with different type policies

Author: This should be fixed now

>   B) Implementation of an interface is added to an existing type which already implements an interface with a different type policy

Reviewer 2: ```suggestion
In particular, this can happen if:

- A type is added to your schema after code generation execution and that type implements two different interfaces with different type policies.
- Implementation of an interface is added to an existing type that already implements an interface with a different type policy.
```

> To learn more, read about how the cache [normalizes objects by cache key](./introduction#normalizing-objects-by-cache-key).

> For most uses cases, you can configure cache keys declaratively using the methods described here. For advanced use cases not supported by declarative cache keys, you can configure cache keys [programmatically](./programmatic-cache-keys).

Reviewer 3: Might just be me but clicking "programmatic cache keys" in the preview makes a 404 (link is https://www.apollographql.com/docs/deploy-preview/1f96ea7a044024b699863083/ios/caching/programmatic-cache-keys)

## Schema Extensions

Declarative cache keys work by adding directives to your backend schema types to indicate how cache keys should be resolved for those types. To do this, you can extend your backend schema by creating a schema extension file with the file extension `.graphqls`. You can add any number of schema extension files to your project.

> **Note:** Make sure that your schema extension files are included in the `schemaSearchPaths` of your code generation configuration's [`FileInput`](./../code-generation/codegen-configuration#file-input).

## The `@typePolicy` directive

The `@typePolicy` directive enables you to specify an object's cache ID from **key fields** of the object returned by your GraphQL server. Type policies can be applied to both concrete object types and interface types in your schema.

To declare a type policy, add an extension to a type which applies the `@typePolicy` directive in a schema extension file.

### Key Fields

The `@typePolicy` directive has a single `keyFields` argument, which takes a string indicating the fields used to determine a type's cache key.

```graphql title="typePolicies.graphqls"
extend type Book @typePolicy(keyFields: "id")
```

> Apollo iOS will now resolve the cache key for all objects with a `__typename` of `"Book"` by using the value of their `id` field. The concrete type of the object is always prepended to the cache key. This means that a `Book` object with an `id` of `456` will resolve to have a cache key of `Book:456`.

Reviewer 3: Might be worth adding that either:

1. it's the user responsibility to make sure all the key fields are queried
2. or that the codegen is doing it automatically

I think it is 1. right now? Either case, it's an important consideration for the system to work.

> **Note:** All of a type's key fields must return a scalar type. To use non-scalar fields in cache keys, try using [programmatic cache key configuration](./programmatic-cache-keys).

### Multiple Key Fields

You can specify _multiple_ key fields for an object if they are all required to uniquely identify a particular cache entry. Multiple key fields should be seperated by a single space.

```graphql title="typePolicies.graphqls"
extend type Author @typePolicy(keyFields: "firstName lastName")
```

The resolved cache key will include all declared key fields separated by the `+` character. In this case, an `Author` object with a `firstName` of `"Ray"` and a last name of `"Bradbury"` would resolve to have a cache key of `Author:Ray+Bradbury`.

### Type policies on interface types

It is also possible to use `@typePolicy` on _interfaces_. This specifies a default set of key fields for all types that implement that interface.

```graphql title="extra.graphqls"
extend interface Node @typePolicy(keyFields: "id")
```

> > **Note:** If an object type has a type policy of its own, it will take precedence and the type policy on an interface it implements will not be applied.

Reviewer: The iOS Codegen will not do this, but will display a validation error instead.

Author: Thanks for the review! I misremembered this. I'll update the docs now.

## Caveats & limitations

> Declarative cache key resolution has a few notable quirks and limitations you should be aware of while implementing your `@typePolicy` directives:

Reviewer 3: Nit, I'd keep that sentence more factual. Something might be a notable quirk to some users but perfectly fine to some others:

```suggestion
Declarative cache key resolution has a few limitations you should be aware of while implementing your `@typePolicy` directives:
```
